### PR TITLE
Fix false positive 'duplicate key' error

### PR DESCRIPTION
**Pull request by evidolob**
This fix for https://github.com/redhat-developer/vscode-yaml/issues/352

For some reasons in `YAMLMapping` keys was only `YAMLScalar` which is wrong, as according to YAML spec, key may be any YAML node. Also detecting of mapping key uniquing  is done for scalar keys, that cause false positive errors `duplicate key` because, in case if key is not scalar, it doesn't have `value` field and code compere two `undefined` values.

This PR fix that, it add compression function, which implemented according to [YAML spec](https://yaml.org/spec/1.2/spec.html#equality//)

### CODE REVIEW

File: test/mappings.test.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Red Hat. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+import * as util from './testUtil';
+
+suite('YAML mappings', () => {
+	test('Flow mappings as keys in mapping', function () {
```
Comment (JPinkney) [inline test/mappings.test.ts]: Looks like theres a bit of a formatting issue here

File: src/yamlAST.ts [hunk 2]
```diff
@@ -127,3 +127,69 @@ export function newMap(mappings?: YAMLMapping[]):YamlMap{
...
     }
 }
+
+/**
+ * Compare nodes as it described in yaml spec https://yaml.org/spec/1.2/spec.html#equality//
+ * 
+ * @param a 
+ * @param b 
+ * 
+ * @returns true when to node equal and false otherwise
```
Comment (JPinkney) [inline src/yamlAST.ts]:
```suggestion
 * @returns true when the two nodes are equal and false otherwise
```

File: test/mappings.test.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    const content = `
+    foo:
+      { bar: baz, foo: fuzz }:
+        - value
+      { qux: quux, bax: bazz }:
+        - value
+    `;
+
+    util.testErrors(content, []);
+	});
```
Comment (JPinkney) [inline test/mappings.test.ts]: I think this another formatting thing. It looks like the document has some tabs in it

File: src/yamlAST.ts [hunk 4]
```diff
@@ -127,3 +127,69 @@ export function newMap(mappings?: YAMLMapping[]):YamlMap{
...
+    if(!a || !b){
+        return false;
+    }
+
+    if(a.kind !== b.kind){
+        return false;
+    }
+
+    if(a.kind === Kind.SCALAR) {
+        return (<YAMLScalar>a).value === (<YAMLScalar>b).value
```
Comment (JPinkney) [inline src/yamlAST.ts]:
```suggestion
        return (<YAMLScalar>a).value === (<YAMLScalar>b).value;
```

**Review by JPinkney — COMMENTED** [review]
I just have a few minor comments

**Review by JPinkney — APPROVED** [review]

File: test/mappings.test.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  test('Flow mappings as keys in mapping', function () {
+    const content = `
+    foo:
+      { bar: baz }:
+        - value
+      { qux: quux }:
+        - value
+    `;
+
+    util.testErrors(content, []);
```
Comment (joshuawilson) [inline test/mappings.test.ts:17]: I'm sure I'm missing something, but I do not see what code this tests. What is the point of these tests?

File: test/yamlAST.test.ts [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+      });
+  
+      test('Scalar Nodes should be equal', () => {
+        const a = newScalar('a');
+        const b = newScalar('a');
+        const result = isNodesEqual(a, b);
+        expect(result).is.true;
+      });
+  
+      test('Scalar Nodes should be not equal', () => {
```
Comment (joshuawilson) [inline test/yamlAST.test.ts]: The text for this should be more clear. It is the opposite of the one above. Reading just the test description is confusing. `Nodes should be equal` `Nodes should not be equal`
Without the condition in the description this is hard to follow. 
May I suggest something like `test('should return not equal when the Scalar Nodes do not equal each other', () => {`

File: test/mappings.test.ts (hunk 5 above)
Comment (JPinkney) [reply test/mappings.test.ts:17]: Previously the content would error with duplicate key because scalars were the only thing allowed. This now checks that flow mapping won't throw errors and that objects can be the values as well. I didn't know this but `util.testErrors(content, []);` automatically asserts inside of it and does the checking. So in this case its basically asserting that no errors are found when it parses

Comment (joshuawilson) [reply test/mappings.test.ts:17]: Interesting, thanks.

**Review by joshuawilson — APPROVED** [review]

**Review by gorkem — APPROVED** [review]
LGTM

**Comment by apisim** [comment]
(By chance) I think I've come across an issue with detecting duplicate keys in mapping nodes and it appears to have been introduced by this #3 PR... Please see a unit test and (potential) fix here: https://github.com/redhat-developer/yaml-ast-parser/compare/master...apimastery:mapping-nodes-dup-key-detection-fix

The unit test is actuall insipred by [a test in yaml-language-server](https://github.com/redhat-developer/yaml-language-server/blob/master/test/schemaValidation.test.ts#L647-L658) - this is how I came across the problem actually and decided to research it.

I am a bit reluctant to create a PR for this because the fix compares only the `key` objects of YAMLMapping nodes so it doesn't really jibe with with the name of the function - `isNodesEqual`. Looking at the current usage of `isNodesEqual`, though, it seems that it is only used to determine if there are duplicates - I didn't find any other places in code in the current `master` branch that call the function other than [this](https://github.com/apimastery/yaml-ast-parser/commit/87a2088a39be8e8530f9342675517a9aa79853e5#diff-1d670561cb1689a37382f052a97a2e97L411-L415) and the unit tests.

Thoughts?

**Comment by evidolob** [comment]
@apisim I'm not sure that understand you, all changes introduced in this PR is not used in yaml-ls yet, as we want to release this project with different name [see PR](https://github.com/redhat-developer/yaml-ast-parser/pull/4).

Can you give an example of yaml where you have issue with detecting duplicate keys? I want to use it in unit test.

**Comment by apisim** [comment]
> @apisim I'm not sure that understand you, all changes introduced in this PR is not used in yaml-ls yet, as we want to release this project with different name [see PR](https://github.com/redhat-developer/yaml-ast-parser/pull/4).
> 
> Can you give an example of yaml where you have issue with detecting duplicate keys? I want to use it in unit test.

Hi @evidolob. Just follow the [link](https://github.com/redhat-developer/yaml-ast-parser/compare/master...apimastery:mapping-nodes-dup-key-detection-fix) - there's a unit test and possible fix. Add the test first and run it; then add the code to yamlAST.ts to fix it. As I said in my comment above, the unit test is actually based on one from yaml-language-server.

I understand that the latest code changes in this project haven't been released yet. A link on npmjs.com for yaml-ast-parser-custom-tags brought me here, so I assumed this will be the replacement for yaml-ast-parser-custom-tags. I was just playing with YAML non-compliant string values (backtick quoted multi-line strings that don't require indentation like when using chomping indicators) and propagating the changes and newly-built custom dependencies to yaml-language-server and the monaco-yaml editor...

I'd suggest to investigate the problem if this is to replace yaml-ast-parser-custom-tags in yaml-language-server - the [unit test here](https://github.com/redhat-developer/yaml-language-server/blob/0.11.1/test/schemaValidation.test.ts#L647-L657) will fail.

Cheers!

**Comment by evidolob** [comment]
@apisim  OK, thx, I will look
